clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs

```diff
@@ -0,0 +1,387 @@
+maybeStringFlag nm opt = ClashFlag
+  { cfFlags = \opts ->
+         strSuffixId nm opts (set opt . Just)
+      ++ unsetArg nm opts (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts
+      then []
+      else case opts ^. opt of
+        Nothing -> renderNoArg nm False
+        Just s -> renderSepSuffix nm s
+  }
+
+
+-- | Return three flags setting a 'Maybe' 'String'. If no argument is supplied, a
+-- default value is set. For example, if the default is @simpleDefault@, then:
+--
+--   * @-fclash-foo@ sets the option to @Just "simpleDefault"@
+--   * @-fclash-foo=wibble@ sets the option to @Just "wibble"@
+--   * @-fclash-no-foo@ sets the option to @Nothing@
+--
+maybeStringWithDefaultFlag ::
+  FlagName ->
+  -- | Default
+  String ->
+  Lens' ClashOpts (Maybe String) ->
+  ClashFlag
+maybeStringWithDefaultFlag flagName dflt opt = ClashFlag
+  { cfFlags = \opts ->
+         anySuffix flagName opts parser (set opt)
+      ++ unsetArg flagName opts (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts
+      then []
+      else case opts ^. opt of
+        Nothing -> renderNoArg flagName False
+        Just s | s == dflt -> renderSuffix flagName ""
+        Just s -> renderSuffix flagName s
+  }
+ where
+  parser :: String -> EwM IO (Maybe (Maybe String))
+  parser ['='] = pure (Just (Just dflt))
+  parser ('=':v) = pure (Just (Just v))
+  parser s = do
+    addErr ("Could not parse: " <> s)
+    pure Nothing
+
+maybeIntFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeIntFlag = maybeReadableFlag
+
+maybeWordFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeWordFlag = maybeReadableFlag
+
+maybeReadableFlag ::
+  forall a.
+  (Eq a, Read a, Show a) =>
+  FlagName ->
+  Lens' ClashOpts (Maybe a) ->
+  ClashFlag
+maybeReadableFlag nm opt = ClashFlag
+  { cfFlags = \ref ->
+         strSuffix nm ref parser (set opt)
+      ++ unsetArg nm ref (set opt)
+
+  , cfRender = \opts ->
+      if isDefault opt opts
+      then []
+      else case opts ^. opt of
+        Nothing -> renderNoArg nm False
+        Just a -> renderSepSuffix nm (show a)
+  }
+ where
+  parser :: String -> EwM IO (Maybe (Maybe a))
+  parser s = case readMaybe s of
+    Nothing -> do
+      addErr ("Could not parse flag argument: " <> s)
+      pure Nothing
+    a -> pure (Just a)
+
+
+readableFlag ::
+  forall a.
+  (Eq a, Read a, Show a) =>
+  FlagName ->
+  Lens' ClashOpts a ->
+  ClashFlag
+readableFlag nm opt = ClashFlag
+  { cfFlags = \ref -> strSuffix nm ref parser (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts then [] else renderSepSuffix nm (show (opts ^. opt))
+  }
+ where
+  parser :: String -> EwM IO (Maybe a)
+  parser s = case readMaybe s of
+    Nothing -> addErr ("Could not parse " <> s)  >> pure Nothing
+    a -> pure a
+
+---------------------------
+-- Generic flag builders --
+---------------------------
+
+-- | Helper function....
```

Comment:
Comments like this aren't particularly helpful, _especially_ as haddock comments

Reply:
Yeah it's a `TODO`